### ISSUE TITLE
[BUG] panic: interface conversion: interface {} is string, not map[string]interface {}

### ISSUE THREAD

**Issue by petrprikryl** [docker/compose#11267]
### Description

`docker compose up` causes:
```
panic: interface conversion: interface {} is string, not map[string]interface {}

goroutine 1 [running]:
github.com/compose-spec/compose-go/v2/loader.ApplyExtends({0x2511d08?, 0xc0007e2750}, 0x16?, {0xc0007d7890, 0x14}, 0xc0002cdc70, 0xc000733f00?, {0xc00053c028, 0x1, 0x1})
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/extends.go:44 +0xb47
github.com/compose-spec/compose-go/v2/loader.loadYamlModel.func1({0x1e37c40?, 0xc0007e2c30?}, {0xc00053c028, 0x1, 0x1})
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/loader.go:318 +0x186
github.com/compose-spec/compose-go/v2/loader.loadYamlModel({0x2511d08, 0xc000666fc0}, {{0x0, 0x0}, {0xc0007d7890, 0x14}, {0xc0007e26f0, 0x1, 0x1}, 0x0}, ...)
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/loader.go:345 +0x82c
github.com/compose-spec/compose-go/v2/loader.ApplyExtends({0x2511d08?, 0xc000666fc0}, 0x16?, {0xc0005a24b0, 0x14}, 0xc0002438f0, 0xc000734400?, {0xc00053c528, 0x1, 0x1})
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/extends.go:69 +0x97f
github.com/compose-spec/compose-go/v2/loader.loadYamlModel.func1({0x1e37c40?, 0xc000a10030?}, {0xc00053c528, 0x1, 0x1})
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/loader.go:318 +0x186
github.com/compose-spec/compose-go/v2/loader.loadYamlModel({0x2511d40, 0xc0000d1900}, {{0x0, 0x0}, {0xc0005a24b0, 0x14}, {0xc000573ad0, 0x1, 0x1}, 0xc000573830}, ...)
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/loader.go:345 +0x82c
github.com/compose-spec/compose-go/v2/loader.load({0x2511d40, 0xc0000d1900}, {{0x0, 0x0}, {0xc0005a24b0, 0x14}, {0xc000573ad0, 0x1, 0x1}, 0xc000573830}, ...)
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/loader.go:404 +0x3ae
github.com/compose-spec/compose-go/v2/loader.LoadWithContext({0x2511d40, 0xc0000d1900}, {{0x0, 0x0}, {0xc0005a24b0, 0x14}, {0xc000573ad0, 0x1, 0x1}, 0xc000573830}, ...)
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/loader/loader.go:274 +0x43a
github.com/compose-spec/compose-go/v2/cli.ProjectFromOptions(0xc000630d00)
        github.com/compose-spec/compose-go/v2@v2.0.0-beta.1/cli/options.go:410 +0x58a
github.com/docker/compose/v2/cmd/compose.(*ProjectOptions).ToProject(0xc000313b80, {0x2524040?, 0xc0001540f0?}, {0x36582c0, 0x0, 0x0}, {0xc000549810?, 0x3?, 0x3?})
        github.com/docker/compose/v2/cmd/compose/compose.go:219 +0x295
github.com/docker/compose/v2/cmd/compose.upCommand.(*ProjectOptions).WithServices.func4({0x2511d40, 0xc0000d1900}, {0x36582c0, 0x0, 0x0})
        github.com/docker/compose/v2/cmd/compose/compose.go:149 +0xee
github.com/docker/compose/v2/cmd/compose.upCommand.(*ProjectOptions).WithServices.Adapt.func6({0x2511d40?, 0xc0000d1900?}, 0x2?, {0x36582c0?, 0x1?, 0x0?})
        github.com/docker/compose/v2/cmd/compose/compose.go:111 +0x30
github.com/docker/compose/v2/cmd/compose.upCommand.(*ProjectOptions).WithServices.Adapt.AdaptCmd.func7(0x0?, {0x36582c0, 0x0, 0x0})
        github.com/docker/compose/v2/cmd/compose/compose.go:90 +0x1d8
github.com/docker/compose/v2/cmd/cmdtrace.Setup.wrapRunE.func2(0xc0005cca00?, {0x36582c0?, 0x0?, 0x0?})
        github.com/docker/compose/v2/cmd/cmdtrace/cmd_span.go:84 +0x65
github.com/spf13/cobra.(*Command).execute(0xc000643800, {0xc000497770, 0x0, 0x0})
        github.com/spf13/cobra@v1.8.0/command.go:983 +0xabc
github.com/spf13/cobra.(*Command).ExecuteC(0xc00057e600)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/docker/cli/cli-plugins/plugin.RunPlugin(0x1c058b3?, 0xc000643500, {{0x21aab10, 0x5}, {0x21b44cf, 0xb}, {0x24ee560, 0x13}, {0x0, 0x0}, ...})
        github.com/docker/cli@v25.0.0-beta.1+incompatible/cli-plugins/plugin/plugin.go:51 +0x10e
github.com/docker/cli/cli-plugins/plugin.Run(0x22e4db0, {{0x21aab10, 0x5}, {0x21b44cf, 0xb}, {0x24ee560, 0x13}, {0x0, 0x0}, {0x0, ...}})
        github.com/docker/cli@v25.0.0-beta.1+incompatible/cli-plugins/plugin/plugin.go:64 +0xe5
main.pluginMain()
        github.com/docker/compose/v2/cmd/main.go:37 +0xa5
main.main()
        github.com/docker/compose/v2/cmd/main.go:75 +0x19c
```

### Steps To Reproduce

docker-compose.common.yml
```
services:
  redis:
    image: redis:7

  redis-extended:
    extends: redis
```

docker-compose.yml
```
services:
  redis:
    extends:
      file: docker-compose.common.yml
      service: redis
```

`docker compose up`

### Compose Version

```Text
Docker Compose version v2.24.0-birthday.10

docker-compose version 1.29.2, build 5becea4c
docker-py version: 5.0.0
CPython version: 3.7.10
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```

### Docker Environment

```Text
Client: Docker Engine - Community
 Version:    24.0.7
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.0-birthday.10
    Path:     /home/petr/.docker/cli-plugins/docker-compose

Server:
 Containers: 9
  Running: 5
  Paused: 0
  Stopped: 4
 Images: 26
 Server Version: 24.0.7
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: 8tc9geqoluipe0v78i5erfc0v
  Is Manager: true
  ClusterID: 1s7uijcx5zhzczskpmg09opcw
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 172.16.0.164
  Manager Addresses:
   172.16.0.164:2377
 Runtimes: runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dd1e886e55dd695541fdcd67420c2888645a495
 runc version: v1.1.10-0-g18a0cb0
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-14-generic
 Operating System: Ubuntu 23.10
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.14GiB
 Name: petrprikryl-Z590M
 ID: 5ZWC:LRYN:PIH6:7XZS:N7QD:H3UE:2NNQ:JMMD:EUKG:XLA5:VD4O:DE4K
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Anything else?

It was working fine on compose version 2.13.

**Comment by ndeloof** [comment]
I can confirm a regression, moving to compose-go/v2 this corner case was uncovered by test case, fixed by https://github.com/compose-spec/compose-go/pull/508

**Comment by jmlord** [comment]
A similar error message can be caused by a mix of two syntax in environment variables:

```
    environment:
      - SOME_VAR=true
      - SOME_OTHER_VAR: true
```

Such a mismatch gives 

> $ docker compose up
> panic: interface conversion: interface {} is map[string]interface {}, not string
> 
> goroutine 1 [running]:
> github.com/compose-spec/compose-go/v2/override.keyValueIndexer({0x1eb8f40?, 0xc00037e6f0?}, {0xc0002c129c?, 0x1?})
>         github.com/compose-spec/compose-go/v2@v2.0.0-rc.2/override/uncity.go:88 +0x79
> ...

**Comment by ArcticFall** [comment]
> A similar error message can be caused by a mix of two syntax in environment variables:
> 
> ```
>     environment:
>       - SOME_VAR=true
>       - SOME_OTHER_VAR: true
> ```
> 
> Such a mismatch gives
> 
> > $ docker compose up
> > panic: interface conversion: interface {} is map[string]interface {}, not string
> > goroutine 1 [running]:
> > github.com/compose-spec/compose-go/v2/override.keyValueIndexer({0x1eb8f40?, 0xc00037e6f0?}, {0xc0002c129c?, 0x1?})
> > github.com/compose-spec/compose-go/v2@v2.0.0-rc.2/override/uncity.go:88 +0x79
> > ...
> 
> (In case someone else lands here with this issue, like I did.)

To clarify, because I was confused by your example at first: This happens if you have used mappings but still have dashes before the entries like in a list.

```
environment:
  - MYSQL_DATABASE: nextcloud
  - MYSQL_INITDB_SKIP_TZINFO: 1
```

Either use only mappings:

```
environment:
  MYSQL_DATABASE: nextcloud
  MYSQL_INITDB_SKIP_TZINFO: 1
```

or only a list:

```
environment:
  - MYSQL_DATABASE=nextcloud
  - MYSQL_INITDB_SKIP_TZINFO=1
```

**Comment by ndeloof** [comment]
```yaml
    environment:
      - SOME_VAR=true
      - SOME_OTHER_VAR: true
```
This just is invalid. `environment` must be either a sequence of strings, using `key(=value)` syntax, or a mapping, but "sequence with mapping" is not supported. Your editor should warn you if it supports schema validation.

**Comment by jmlord** [comment]
> This just is invalid. `environment` must be either a sequence of strings, using `key(=value)` syntax, or a mapping, but "sequence with mapping" is not supported. Your editor should warn you if it supports schema validation.

Clearly it is, and clearly schema validation should warn us, but it seems I didn't have a Visual Studio plugin doing it in that instance. And when the code actually has a bunch of comments between each env var, it can easily go unnoticed. 

However, beyond the code editor, shouldn't docker compose give us a more convenient message then `panic: interface conversion: interface {} is map[string]interface {}, not string` ? A line number could help too... In fact, a compiler-style message would be much better. Something like:
```
String expected as environment at line 63 of path/to/compose.yml.
Map[string] was found instead.
<stack trace>
```

**Comment by ndeloof** [comment]
> Clearly it is

Definitively **NO**. `- SOME_OTHER_VAR: true` is a list entry with a nested structure. Whenever it only has one line. This is not equivalent to `SOME_OTHER_VAR: true` (without `-`).

**Comment by jmlord** [comment]
Clearly it is [invalid] ... :)

**Comment by ndeloof** [comment]
sorry my bad. Answering too much issues in parallel, so not reading carefully enough

**Comment by faizal108** [comment]
this cause an error : 
if you use dash before key then use equal(=) instead colon(:)
![image](https://github.com/user-attachments/assets/fad4ebcf-f6b7-4d59-a518-c4c6ec38cbb7)

Correct way : 
![image](https://github.com/user-attachments/assets/a6f9e447-ad79-4622-966a-3360b660a644)
or 
![image](https://github.com/user-attachments/assets/61bda92c-e1c1-408a-9db3-b2e17329eb9f)

**Comment by rismanma1** [comment]
I encountered this today, when I was splitting one of my compose files into two. Unfortunately I missed copying the first line, "services:", so when running docker compose, I encountered "panic: interface conversion: interface {} is nil, not map[string]interface {}", accompanied by its stack trace. Here was my file:
```
  bos11:
    extends:
      file: docker-compose-common.yaml
      service: bos1
    networks:
      default:
        ipv4_address: 192.168.11.31
``` 
Does this add anything useful? I could include more details but otherwise I don't want to clutter things up here more than necessary
